modules/baremetal/main.tf

```diff
@@ -0,0 +1,36 @@
+resource "ibm_is_bare_metal_server" "bms" {
+  profile     = var.profile
+  name        = var.prefix
+  image       = var.image
+  zone        = var.zone
+  keys        = var.ssh_key_id
+  vpc         = var.vpc_id
+  bandwidth   = var.bandwidth
+  access_tags = var.access_tags
+
+  # Use primary_network_interface if no VLANs are needed
+  dynamic "primary_network_interface" {
+    for_each = length(var.allowed_vlans) == 0 ? [1] : []
+    content {
+      subnet = var.subnet_id
+    }
+  }
+
+  # Use primary_network_attachment if VLANs are required
+  dynamic "primary_network_attachment" {
+    for_each = length(var.allowed_vlans) > 0 ? [1] : []
+    content {
+      name = "${var.prefix}-vni"
+      virtual_network_interface {
+        id = ibm_is_virtual_network_interface.bms[0].id
```

Comment:
I think there is a logic issue here somewhere, the `ibm_is_virtual_network_interface.bms` will end up with multiple resources, one per subnet in the given subnet list. But the only place I see them used is here, and you are hard-coding to only use the first one ( [0] ). I'm not sure what was the intent here, but you are potentially creating multiple VNIs that will never be used or attached?